src/Datadog.Trace/SpanContextPropagator.cs

```diff
@@ -0,0 +1,124 @@
+using System;
+using System.Globalization;
+using System.Linq;
+using Datadog.Trace.Headers;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace
+{
+    internal class SpanContextPropagator
+    {
+        private const NumberStyles NumberStyles = System.Globalization.NumberStyles.Integer;
+
+        private static readonly CultureInfo InvariantCulture = CultureInfo.InvariantCulture;
+        private static readonly ILog Log = LogProvider.For<SpanContextPropagator>();
+
+        private SpanContextPropagator()
+        {
+        }
+
+        public static SpanContextPropagator Instance { get; } = new SpanContextPropagator();
+
+        /// <summary>
+        /// Propagates the specified context by adding new headers to a <see cref="IHeadersCollection"/>.
+        /// This locks the sampling priority for <paramref name="context"/>.
+        /// </summary>
+        /// <param name="context">A <see cref="SpanContext"/> value that will be propagated into <paramref name="headers"/>.</param>
+        /// <param name="headers">A <see cref="IHeadersCollection"/> to add new headers to.</param>
+        public void Inject(SpanContext context, IHeadersCollection headers)
+        {
+            if (context == null) { throw new ArgumentNullException(nameof(context)); }
+
+            if (headers == null) { throw new ArgumentNullException(nameof(headers)); }
+
+            headers.Set(HttpHeaderNames.TraceId, context.TraceId.ToString(InvariantCulture));
+            headers.Set(HttpHeaderNames.ParentId, context.ParentId?.ToString(InvariantCulture));
+
+            var traceContext = context.TraceContext;
+            var samplingPriority = (int?)traceContext.SamplingPriority;
+
+            headers.Set(
+                HttpHeaderNames.SamplingPriority,
+                samplingPriority?.ToString(InvariantCulture));
+
+            // lock sampling priority when span propagates.
+            // if sampling priority is not set yet, this will determine
+            // a value using a Sampler.
+            traceContext.LockSamplingPriority();
```

Comment:
We should lock before setting the header, since if we don't have one here, and then we roll the dice, we'll want to propagate whatever we roll to the next service.

Reply:
Oops! Fixed in f190068ee71e27ada6342bac16a350ab2fa48f37.